File: matlab/+pulserver/report.m
function info = report(seq, varargin)
% report - Structured report of the sequence collection.
%
%   info = pulserver.report(seq)
%   str  = pulserver.report(seq, 'print', true)
%
% Inputs
%   seq     double | SequenceCollection   Collection handle or object.
%
% Name-value options
%   print   logical     If true, return a formatted string. Default: false
%
% Output
%   info    struct array (1 x num_subsequences) with fields:
%       tr_size              - blocks per TR
%       num_trs              - total TRs
%       tr_duration_us       - TR duration (us)
%       num_prep_blocks      - preparation blocks
%       num_cooldown_blocks  - cooldown blocks
%       num_unique_segments  - unique segments
%       segments             - struct array of (start_block, num_blocks)
%       prep_segment_table   - segment IDs for prep region
%       main_segment_table   - segment IDs for main TR
%       cooldown_segment_table - segment IDs for cooldown
%
% See also: pulserver.SequenceCollection, pulserver.get_block

    p = inputParser;
    addParameter(p, 'print', false, @islogical);
    parse(p, varargin{:});

    info = pulseqlib_mex('report', to_handle(seq));

    if p.Results.print
        nss = numel(info);
        lines = {};
        lines{end+1} = sprintf('Subsequences: %d', nss);
        lines{end+1} = '';
        for ss = 1:nss
            s = info(ss);
            lines{end+1} = sprintf('--- Subsequence %d ---', ss);
            lines{end+1} = sprintf('  TR size:            %d blocks', s.tr_size);
            lines{end+1} = sprintf('  TR duration:        %.3f ms', s.tr_duration_us / 1e3);
            lines{end+1} = sprintf('  Prep blocks:        %d', s.num_prep_blocks);
            lines{end+1} = sprintf('  Cooldown blocks:    %d', s.num_cooldown_blocks);
            lines{end+1} = sprintf('  Unique segments:    %d', s.num_unique_segments);
            lines{end+1} = sprintf('  Segment order (TR): [%s]', ...
                            num2str(s.main_segment_table));
            for si = 1:numel(s.segments)
                lines{end+1} = sprintf('    seg %d: start_block=%d, num_blocks=%d', ...
                    si, s.segments(si).start_block, s.segments(si).num_blocks);
            end
            lines{end+1} = '';
        end
        info = strjoin(lines, newline);
    end
end

function h = to_handle(seq)
    if isa(seq, 'pulserver.SequenceCollection')
        h = seq.Handle;
    else
        h = seq;
    end
end

File: python/pulserver/core/_report.py
"""Sequence collection report for SequenceCollection."""

__all__ = ['report']

from types import SimpleNamespace

from ._extension._pulseqlib_wrapper import _get_report
from ._sequence import SequenceCollection


def report(
    seq: SequenceCollection,
    *,
    do_print: bool = False,
) -> list[SimpleNamespace]:
    """Generate a structured report of the sequence collection.

    Returns one :class:`~types.SimpleNamespace` per subsequence with:

    - ``unique_block_ids`` — list of (segment_idx, block_idx) for the
      first instance of each unique block in the sequence.
    - ``segments`` — list of ``(start_block, num_blocks)`` tuples, where
      ``start_block`` is the 0-based index in the original Sequence of
      the segment instance with maximum energy.
    - ``num_prep_blocks`` — preparation blocks before first TR.
    - ``num_cooldown_blocks`` — cooldown blocks after last TR.
    - ``tr_size`` — number of blocks per TR.
    - ``tr_duration_s`` — TR duration in seconds.
    - ``segment_order`` — ordered list of segment IDs composing one TR.

    Parameters
    ----------
    seq : SequenceCollection
        The sequence to report on.
    do_print : bool
        If ``True``, return a formatted string instead of a list of
        SimpleNamespaces.

    Returns
    -------
    list[SimpleNamespace] or str
        One entry per subsequence (or a formatted string if *do_print*).
    """
    raw = _get_report(seq._cseq)

    num_ss = raw["num_subsequences"]
    results = []

    for ss_dict in raw["subsequences"]:
        ns = SimpleNamespace()

        # -- unique block IDs (segment_idx, block_idx) --
        block_ids = []
        seg_offset = ss_dict["segment_offset"]
        for i, seg in enumerate(ss_dict["segments"]):
            for blk in range(seg["num_blocks"]):
                block_ids.append((seg_offset + i, blk))
        ns.unique_block_ids = block_ids

        # -- segments: (start_block, num_blocks) --
        ns.segments = [
            (seg["start_block"], seg["num_blocks"])
            for seg in ss_dict["segments"]
        ]

        ns.num_prep_blocks = ss_dict["num_prep_blocks"]
        ns.num_cooldown_blocks = ss_dict["num_cooldown_blocks"]
        ns.tr_size = ss_dict["tr_size"]
        ns.tr_duration_s = ss_dict["tr_duration_us"] * 1e-6

        # -- ordered segment IDs composing one TR --
        ns.segment_order = list(ss_dict["main_segment_table"])

        # -- extra tables --
        ns.prep_segment_table = list(ss_dict["prep_segment_table"])
        ns.cooldown_segment_table = list(ss_dict["cooldown_segment_table"])

        results.append(ns)

    if not do_print:
        return results

    # ── formatted output ──────────────────────────────────────────
    lines = []
    total_dur_s = raw["total_duration_us"] * 1e-6
    lines.append(f"Sequence length: {total_dur_s:.6f} s  |  "
                 f"Subsequences: {num_ss}  |  "
                 f"Total unique segments: {raw['num_segments']}")
    lines.append("")

    for idx, ns in enumerate(results):
        lines.append(f"--- Subsequence {idx} ---")
        lines.append(f"  TR size:            {ns.tr_size} blocks")
        lines.append(f"  TR duration:        {ns.tr_duration_s * 1e3:.3f} ms")
        lines.append(f"  Prep blocks:        {ns.num_prep_blocks}")
        lines.append(f"  Cooldown blocks:    {ns.num_cooldown_blocks}")
        lines.append(f"  Unique blocks:      {len(ns.unique_block_ids)}")
        lines.append(f"  Unique segments:    {len(ns.segments)}")
        lines.append(f"  Segment order (TR): {ns.segment_order}")
        for si, (start, nblk) in enumerate(ns.segments):
            lines.append(f"    seg {si}: start_block={start}, "
                         f"num_blocks={nblk}")
        lines.append("")

    return "\n".join(lines)
